Fix missing null terminator in tools/mkfs.c

olivia-fl opened the pull request:
strncpy does not write a null terminator if the src string is longer than the dst size. If a user passed a filename that was longer than DIRSIZ-1, then we would silently write a non-null-terminated string to `de.d_name`.

This bug triggers a warning on GCC 14.2.1:

```
tools/mkfs.c: In function ‘main’:
tools/mkfs.c:254:17: error: ‘__builtin_strncpy’ specified bound 254 equals destination size [-Werror=stringop-truncation]
  254 |                 strncpy(de.d_name, name, DIRSIZ);
      |                 ^
```

olivia-fl committed:
fix missing null terminator in tools/mkfs.c

strncpy does not write a null terminator if the src string is longer
than the dst size. If a user passed a filename that was longer than
DIRSIZ-1, then we would silently write a non-null-terminated string to
de.d_name. Instead, truncate at DIRSIZ-1, write a null terminator, and
emit a warning.

This bug triggered a warning on GCC 14.2.1:

tools/mkfs.c: In function ‘main’:
tools/mkfs.c:254:17: error: ‘__builtin_strncpy’ specified bound 254 equals destination size [-Werror=stringop-truncation]
  254 |                 strncpy(de.d_name, name, DIRSIZ);
      |                 ^

Connor-GH reviewed (changes requested):
Remove the ``return 1;`` and change the error message to say something like ``WARNING: filename being truncated: `%s'`` please.

Connor-GH commented:
Also for the record, according to https://en.wikipedia.org/wiki/Comparison_of_file_systems, the only filesystems that can handle file sizes larger than 255 are HAMMER (DragonflyBSD) and ZFS (Linux/*BSD).

Connor-GH commented:
Small nit: please change the ``= 0;`` to be ``= '\0';``. Last thing and then it's ready to be merged.

Connor-GH commented:
In all fairness the code style guidelines are a working draft ATM.

olivia-fl commented:
Switched to a warning, and changed strncpy so that it has a null terminator even when the string is truncated.

> Also for the record, according to https://en.wikipedia.org/wiki/Comparison_of_file_systems, the only filesystems that can handle file sizes larger than 255 are HAMMER (DragonflyBSD) and ZFS (Linux/*BSD).

I think I'm a bit confused, the names are coming from argv right? Are they limited by what's representable in the filesystem?

Connor-GH commented:
> I think I'm a bit confused, the names are coming from argv right? Are they limited by what's representable in the filesystem?

The names being copied are names of files on disk. Technically i *still* do this wrong because this is supposed to be 255 bytes per file, but it is instead 254 bytes (FS limit) per file string from the root of the git directory. I will cross that bridge when I get to it.
